### DISCUSSION TITLE
use remix with UI libraries

### DISCUSSION THREAD

**Discussion by iwan-uschka** [remix-run/remix#3507] [Q&A] [answered]
I just tried to create a little remix app using a UI library called "React Suite" (https://rsuitejs.com/). I am used to this library so i thought i gonna use it in this project. But i got stuck pretty quick when creating forms. `@remix-run/react` provides a `Form` component, which does all the magic regarding data transfer. `rsuite` on the other hand provides a `Form` component, which does all the magic regarding UI (layout, design, validation). Unfortunately it doesn't seem to be possible to combine those two components. All the subcomponents of `rsuite`s `Form` component (https://rsuitejs.com/components/form/) rely on that component

```
<Form>
    <Form.Group controlId="name">
      <Form.ControlLabel>Username</Form.ControlLabel>
      <Form.Control name="name" />
      <Form.HelpText>Username is required</Form.HelpText>
    </Form.Group>
    ...
    <Form.Group>
      <ButtonToolbar>
        <Button appearance="primary">Submit</Button>
        <Button appearance="default">Cancel</Button>
      </ButtonToolbar>
    </Form.Group>
</Form>
```

so i cannot use `@remix-run/react`'s `Form` component here as a wrapper. Unfortunately.

So 2 questions:
1. Is it planned to add an `as` prop to `@remix-run/react`'s `Form` component so the `form` tag is replaced by any component given via this prop?
2. Is there any "easy" way to do all the data transfer magic which `@remix-run/react`'s `Form` does without using this component and without reimplementing the magic? `rsuite`s `Form` component provides an `onSubmit` callback.

**Comment by kiliman** [comment, marked as answer]
Does this library render standard HTML inputs? That is does it submit like a traditional form or do you have to add an onSubmit handler?

Also Remix has a useSubmit hook that is a programmatic way to submit a form just like the Form component. 

https://remix.run/docs/en/v1/api/remix#usesubmit

**Reply by iwan-uschka** [reply]
Exactly that did the trick! Thanks @kiliman !!

**Comment by CibiAananth** [comment]
@iwan-uschka I am looking to setup rsuite with remix. Do you have any public repo that has the code or any reference docs?

**Reply by iwan-uschka** [reply]
Here is what i came up with (more or less):

```
import { useTransition, useSubmit } from "@remix-run/react";

import {
  Form as RsForm,
  Button as RsButton,
  ButtonToolbar as RsButtonToolbar,
  Schema as RsSchema
} from "rsuite";

const ComponentWithForm = () => {
  const submit = useSubmit();
  const transition = useTransition();

  return (
    <RsForm
      onSubmit={(isFormValid, event) => {
        if (isFormValid) {
          submit(event.currentTarget);
        }
      }}
      model={RsSchema.Model({
        formField1: RsSchema.Types.StringType().isRequired(
          "This field is required."
        ),
        formField2: RsSchema.Types.StringType().isRequired(
          "This field is required."
        ),
      })}
    >
      <RsForm.Group>
        <RsForm.ControlLabel>form field 1</RsForm.ControlLabel>
        <RsForm.Control name="formField1" />
      </RsForm.Group>
      <RsForm.Group>
        <RsForm.ControlLabel>form field 2</RsForm.ControlLabel>
        <RsForm.Control name="formField2" />
      </RsForm.Group>
      <RsForm.Group>
        <RsButtonToolbar>
          <RsButton
            appearance="primary"
            type="submit"
            loading={transition.state === "submitting"}
          >
            submit
          </RsButton>
        </RsButtonToolbar>
      </RsForm.Group>
    </RsForm>
  );
};

export default ComponentWithForm;
```

This code hasn't been touched for a long time so probably it could be outdated. Dependencies:
- remix@1.3.4
- rsuite@5.15.1

**Comment by CibiAananth** [comment]
@iwan-uschka Thanks will check this.
